testhelpers/arg_patterns.go

```diff
@@ -0,0 +1,83 @@
+package testhelpers
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/buildpacks/pack/internal/stringset"
+)
+
+func AssertIncludeAllExpectedPatterns(t *testing.T, receivedArgs []string, expectedPatterns ...[]string) {
+	missingPatterns := [][]string{}
+
+	for _, expectedPattern := range expectedPatterns {
+		if !patternExists(expectedPattern, receivedArgs) {
+			missingPatterns = append(missingPatterns, expectedPattern)
+		}
+	}
+
+	assertSliceEmpty(t,
+		missingPatterns,
+		"Expected the patterns %s to exist in [%s]",
+		missingPatterns,
+		strings.Join(receivedArgs, " "),
+	)
+}
+
+func patternExists(expectedPattern []string, receivedArgs []string) bool {
+	_, missing, _ := stringset.Compare(receivedArgs, expectedPattern)
+	if len(missing) > 0 {
+		return false
+	}
+
+	if len(expectedPattern) == 1 {
+		return true
+	}
+
+	for _, loc := range matchLocations(expectedPattern[0], receivedArgs) {
+		finalElementLoc := loc + len(expectedPattern)
+
+		receivedSubSlice := receivedArgs[loc:finalElementLoc]
+
+		if reflect.DeepEqual(receivedSubSlice, expectedPattern) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchLocations(expectedArg string, receivedArgs []string) []int {
+	indices := []int{}
+
+	for i, receivedArg := range receivedArgs {
+		if receivedArg == expectedArg {
+			indices = append(indices, i)
+		}
+	}
+
+	return indices
+}
+
+func assertSliceEmpty(t *testing.T, actual interface{}, msg string, msgArgs ...interface{}) {
+	empty, err := sliceEmpty(actual)
+
+	if err != nil {
+		t.Fatalf("assertSliceNotEmpty error: %s", err.Error())
+	}
+
+	if !empty {
+		t.Fatalf(msg, msgArgs...)
+	}
+}
+
+func sliceEmpty(slice interface{}) (bool, error) {
+	switch reflect.TypeOf(slice).Kind() {
+	case reflect.Slice:
+		return reflect.ValueOf(slice).Len() == 0, nil
+	default:
+		return true, fmt.Errorf("invoked with non slice actual: %v", slice)
+	}
+}
```

Comment:
I know I added these methods, but I'm wondering if they need to be as flexible now. Do we ever pass these anything other than a slice of string? If we don't need to be generic as to what sort of slice gets passed we shouldn't.

Reply:
You are right, we only ever pass string slices. Making this change.